File: output/xCodeEval_remove_8000/C++/Java/29d33ec3efff45ee9f12b18e7461d439.java
import java.util.*;
import java.io.*;

class Main {
    static class Pair {
        int first, second;
        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }

    static class PairPair {
        Pair first, second;
        PairPair(Pair first, Pair second) {
            this.first = first;
            this.second = second;
        }
    }

    static final int Mod = 1000000007, INF = 1e9, LINF = 1e18;
    static final double Pi = 3.141592653589793116, EPS = 1e-9, Gold = ((1+Math.sqrt(5))/2);
    static final long[] keymod = {1000000007, 1000000009, 1000000021, 1000000033};
    static final int keyCount = keymod.length;

    static int getbit(long s, int i) { return (int)((s >> i) & 1); }
    static long onbit(long s, int i) { return s | (1L << i); }
    static long offbit(long s, int i) { return s & (~(1L << i)); }
    static int cntbit(long s) { return Long.bitCount(s); }

    static long TimeStart, TimeEnd;

    static void ControlIO(String[] args) {
        String infile = "FILE.IN";
        String outfile = "FILE.OUT";

        System.out.println("Source code by ");
        System.out.println("Current time: " + new SimpleDateFormat("yyyy-MM-dd | HH:mm:ss").format(new Date()) + "\n");

        System.out.println("OI-Mode: ON");
        System.out.println("Input file: " + infile);
        System.out.println("Output file: " + outfile + "\n");

        System.out.println("OI-Mode: OFF");
        System.out.println("Input file: NULL");
        System.out.println("Output file: NULL\n");

        System.out.println("MultiTest-Mode: ON");

        System.out.println("MultiTest-Mode: OFF");

        System.out.println("Interactive-Mode: ON\n");

        System.out.println("Interactive-Mode: OFF\n");

        if (args.length > 0) {
            try {
                System.setIn(new FileInputStream(args[0]));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        if (args.length > 1) {
            try {
                System.setOut(new PrintStream(new FileOutputStream(args[1])));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    static void TimerStart() {
        TimeStart = System.nanoTime();
    }

    static void TimerStop() {
        TimeEnd = System.nanoTime();
        double ElapsedTime = (double)(TimeEnd - TimeStart) / 1e9;
        System.out.println("\n\nTime elapsed: " + ElapsedTime + " seconds.");
    }

    static void Exit() {
        TimerStop();
        System.exit(0);
    }

    static String s1, s2, virus;
    static int n1, n2, nv;
    static List<List<Integer>> id1;

    static void ProceedVirus() {
        id1 = new ArrayList<>();
        for (int i = 0; i < nv; i++) {
            id1.add(new ArrayList<>());
            for (int j = 0; j < 26; j++) {
                String pref2 = virus.substring(0, i);
                pref2 += (char)(65 + j);
                for (int x = i + 1; x >= 0; x--) {
                    if (pref2.substring(i + 1 - x, x).equals(virus.substring(0, x))) {
                        id1.get(i).add(x);
                        break;
                    }
                }
            }
        }
    }

    static void Input() {
        Scanner sc = new Scanner(System.in);
        s1 = sc.next();
        s2 = sc.next();
        virus = sc.next();
        nv = virus.length();
        n1 = s1.length();
        n2 = s2.length();
    }

    static void Solve() {
        List<List<Integer>> List1 = new ArrayList<>();
        List<List<Integer>> List2 = new ArrayList<>();
        for (int i = 0; i < 26; i++) {
            List1.add(new ArrayList<>());
            List2.add(new ArrayList<>());
        }
        for (int i = 0; i < n1; i++) {
            List1.get(s1.charAt(i) - 'A').add(i);
        }
        for (int i = 0; i < n2; i++) {
            List2.get(s2.charAt(i) - 'A').add(i);
        }
        ProceedVirus();
        PairPair Default = new PairPair(new Pair(-1, -1), new Pair(-1, -1));
        boolean[][][][] dp = new boolean[n1][n2][Math.min(n1, n2) + 1][nv + 1];
        PairPair[][][][] Last = new PairPair[n1][n2][Math.min(n1, n2) + 1][nv + 1];
        for (int i = 0; i < n1; i++) {
            for (int j = 0; j < n2; j++) {
                if (s1.charAt(i) == s2.charAt(j)) {
                    dp[i][j][1][id1.get(0).get(s1.charAt(i) - 'A')] = true;
                } else {
                    dp[i][j][0][0] = true;
                }
            }
        }
        for (int x = 1; x < Math.min(n1, n2); x++) {
            for (int i = 0; i < n1 - 1; i++) {
                for (int j = 0; j < n2 - 1; j++) {
                    for (int y = 0; y < nv; y++) {
                        if (!dp[i][j][x][y]) continue;
                        for (int z = 0; z < 26; z++) {
                            if (Collections.binarySearch(List1.get(z), i) == List1.get(z).size()) continue;
                            if (Collections.binarySearch(List2.get(z), j) == List2.get(z).size()) continue;
                            int i1 = List1.get(z).get(Collections.binarySearch(List1.get(z), i));
                            int j1 = List2.get(z).get(Collections.binarySearch(List2.get(z), j));
                            dp[i1][j1][x + 1][id1.get(y).get(z)] = true;
                            Last[i1][j1][x + 1][id1.get(y).get(z)] = new PairPair(new Pair(i, j), new Pair(x, y));
                        }
                    }
                }
            }
        }
        for (int x = Math.min(n1, n2); x > 0; x--) {
            for (int y = 0; y < nv; y++) {
                for (int i = 0; i < n1; i++) {
                    for (int j = 0; j < n2; j++) {
                        if (!dp[i][j][x][y]) continue;
                        int I = i, J = j, X = x, Y = y;
                        StringBuilder s = new StringBuilder();
                        while (I != -1) {
                            s.insert(0, s1.charAt(I));
                            PairPair Token = Last[I][J][X][Y];
                            I = Token.first.first;
                            J = Token.first.second;
                            X = Token.second.first;
                            Y = Token.second.second;
                        }
                        System.out.println(s);
                        return;
                    }
                }
            }
        }
        System.out.println("0");
    }

    public static void main(String[] args) {
        ControlIO(args);

        Input();
        TimerStart();
        Solve();
        TimerStop();

        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        TimerStart();
        while (T-- > 0) {
            Input();
            Solve();
        }
        TimerStop();
    }
}

File: output/xCodeEval_remove_8000/C++/C/29d33ec3efff45ee9f12b18e7461d439.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_SIZE 1000000

typedef long long i64;
typedef long double ld;

typedef struct {
    i64 first;
    i64 second;
} pii;

typedef struct {
    i64 first;
    pii second;
} pip;

typedef struct {
    pii first;
    i64 second;
} ppi;

const i64 Mod = 1000000007LL, INF = 1e9, LINF = 1e18;
const ld Pi = 3.141592653589793116, EPS = 1e-9, Gold = ((1+sqrt(5))/2);
i64 keymod[] = {1000000007LL, 1000000009LL, 1000000021LL, 1000000033LL};
i64 keyCount = sizeof(keymod) / sizeof(i64);

typedef struct {
    i64 s[MAX_SIZE];
    int top;
} Stack;

void init(Stack *stack) {
    stack->top = -1;
}

void push(Stack *stack, i64 value) {
    stack->s[++stack->top] = value;
}

i64 pop(Stack *stack) {
    return stack->s[stack->top--];
}

i64 getbit(i64 s, int i) {
    return (s >> i) & 1;
}

i64 onbit(i64 s, int i) {
    return s | (1LL << i);
}

i64 offbit(i64 s, int i) {
    return s & (~(1LL << i));
}

int cntbit(i64 s) {
    int count = 0;
    while (s) {
        count += s & 1;
        s >>= 1;
    }
    return count;
}

void proceedVirus(i64 id1[][26], i64 nv, char *virus) {
    for (i64 i = 0; i < nv; i++) {
        for (i64 j = 0; j < 26; j++) {
            char pref2[MAX_SIZE];
            strncpy(pref2, virus, i);
            pref2[i] = 'A' + j;
            pref2[i+1] = '\0';
            for (i64 x = i + 1; x >= 0; x--) {
                if (strncmp(pref2 + i + 1 - x, virus, x) == 0) {
                    id1[i][j] = x;
                    break;
                }
            }
        }
    }
}

void input(char *s1, char *s2, char *virus, i64 *n1, i64 *n2, i64 *nv) {
    scanf("%s %s %s", s1, s2, virus);
    *n1 = strlen(s1);
    *n2 = strlen(s2);
    *nv = strlen(virus);
}

void solve(char *s1, char *s2, char *virus, i64 n1, i64 n2, i64 nv) {
    i64 id1[MAX_SIZE][26];
    proceedVirus(id1, nv, virus);
    
    i64 List1[26][MAX_SIZE], List2[26][MAX_SIZE];
    i64 List1Size[26] = {0}, List2Size[26] = {0};
    for (i64 i = 0; i < n1; i++) {
        List1[s1[i]-'A'][List1Size[s1[i]-'A']++] = i;
    }
    for (i64 i = 0; i < n2; i++) {
        List2[s2[i]-'A'][List2Size[s2[i]-'A']++] = i;
    }
    
    i64 dp[MAX_SIZE][MAX_SIZE][MAX_SIZE][MAX_SIZE] = {0};
    pii Default = {-1, -1};
    ppp Last[MAX_SIZE][MAX_SIZE][MAX_SIZE][MAX_SIZE];
    for (i64 i = 0; i < n1; i++) {
        for (i64 j = 0; j < n2; j++) {
            if (s1[i] == s2[j]) {
                dp[i][j][1][id1[0][s1[i]-'A']] = 1;
            } else {
                dp[i][j][0][0] = 1;
            }
        }
    }
    for (i64 x = 1; x < (n1 < n2 ? n1 : n2); x++) {
        for (i64 i = 0; i < n1 - 1; i++) {
            for (i64 j = 0; j < n2 - 1; j++) {
                for (i64 y = 0; y < nv; y++) {
                    if (!dp[i][j][x][y]) {
                        continue;
                    }
                    for (i64 z = 0; z < 26; z++) {
                        if (List1Size[z] == 0 || List2Size[z] == 0) {
                            continue;
                        }
                        i64 i1 = -1, j1 = -1;
                        for (i64 k = 0; k < List1Size[z]; k++) {
                            if (List1[z][k] > i) {
                                i1 = List1[z][k];
                                break;
                            }
                        }
                        for (i64 k = 0; k < List2Size[z]; k++) {
                            if (List2[z][k] > j) {
                                j1 = List2[z][k];
                                break;
                            }
                        }
                        if (i1 == -1 || j1 == -1) {
                            continue;
                        }
                        dp[i1][j1][x+1][id1[y][z]] = 1;
                        Last[i1][j1][x+1][id1[y][z]].first.first = i;
                        Last[i1][j1][x+1][id1[y][z]].first.second = j;
                        Last[i1][j1][x+1][id1[y][z]].second.first = x;
                        Last[i1][j1][x+1][id1[y][z]].second.second = y;
                    }
                }
            }
        }
    }
    for (i64 x = (n1 < n2 ? n1 : n2); x > 0; x--) {
        for (i64 y = 0; y < nv; y++) {
            for (i64 i = 0; i < n1; i++) {
                for (i64 j = 0; j < n2; j++) {
                    if (!dp[i][j][x][y]) {
                        continue;
                    }
                    i64 I = i, J = j, X = x, Y = y;
                    char s[MAX_SIZE];
                    int index = 0;
                    while (I != -1) {
                        s[index++] = s1[I];
                        ppp Token = Last[I][J][X][Y];
                        I = Token.first.first;
                        J = Token.first.second;
                        X = Token.second.first;
                        Y = Token.second.second;
                    }
                    s[index] = '\0';
                    printf("%s\n", s);
                    return;
                }
            }
        }
    }
    printf("0\n");
}

int main() {
    char s1[MAX_SIZE], s2[MAX_SIZE], virus[MAX_SIZE];
    i64 n1, n2, nv;
    input(s1, s2, virus, &n1, &n2, &nv);
    solve(s1, s2, virus, n1, n2, nv);
    return 0;
}
